File: Python/algorithmic_heights/13_par.py
# 2-Way Partition
import sys


def convert_to_intlist(line):
    return [int(item) for item in line.split()]


def two_way_partitioning(array, pivot, n):
    start_ix = 1
    end_ix = n - 1
    while start_ix <= end_ix:
        elem = array[start_ix]
        if elem <= pivot:
            array[start_ix], array[start_ix-1] = array[start_ix-1], elem
            start_ix += 1
        else:
            array[start_ix], array[end_ix] = array[end_ix], elem
            end_ix -= 1


def main():
    reader = sys.stdin
    n = int(next(reader))
    array = convert_to_intlist(next(reader))
    two_way_partitioning(array, array[0], n)
    print(' '.join(map(str, array)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/13_par.hs
import Control.Monad.ST (runST)
import Data.Vector (Vector, freeze, fromList, thaw, toList)
import qualified Data.Vector as V
import qualified Data.Vector.Mutable as MV

convertToIntList :: String -> [Int]
convertToIntList = map read . words


twoWayPartitioning :: Vector Int -> Int -> Int -> Vector Int
twoWayPartitioning vector pivot n =
    let swapElements array =
            let go startIx endIx =
                    if startIx > endIx then return ()
                    else do
                        elem <- MV.read array startIx
                        if elem < pivot then do
                            MV.swap array startIx (startIx - 1)
                            go (startIx + 1) endIx
                        else do
                            MV.swap array startIx endIx
                            go startIx (endIx - 1)
            in go 1 (n - 1)
    in runST $ do
        array <- thaw vector
        swapElements array
        freeze array


main :: IO ()
main = do
    n <- readLn
    xs <- fromList . convertToIntList <$> getLine
    let result = twoWayPartitioning xs (V.head xs) n
    putStrLn $ unwords $ map show $ toList result
